src/Pull.cpp

```diff
@@ -0,0 +1,200 @@
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_connect";
+   } else{
+      connectResponse = nn_bind(mSocket,
+                                mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_bind";
+   }
+   if (connectResponse < 0) {
+      throw std::runtime_error("could not connect to endpoint: " +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   LOG(INFO) << connectionType
+              << " socket at: " << mProtocolHandler->GetLocation()
+              << " with timeout: " << timeoutInMs;
+}
+/**
+ * Construct a non-blocking nanomsg pull (timeout)
+ */
+Pull::Pull(const std::string& location,
+           const int timeoutInMs,
+           const int bufferSize,
+           const bool shouldConnect) {
+   mProtocolHandler = std::move(std::unique_ptr<NanoProtocol>(
+                                   new NanoProtocol(location)));
+   mSocket = nn_socket(AF_SP, NN_PULL);
+   if (mSocket < 0) {
+      throw std::runtime_error("could not open transport socket" +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   if (timeoutInMs > 0) {
+      nn_setsockopt(mSocket,
+                    NN_SOL_SOCKET,
+                    NN_RCVTIMEO,
+                    &timeoutInMs,
+                    sizeof(timeoutInMs));
+   } else {
+      LOG(INFO) << " zero or negative timeout requested, blocking socket";
+   }
+   nn_setsockopt(mSocket,
+                 NN_SOL_SOCKET,
+                 NN_RCVBUF,
+                 &bufferSize,
+                 sizeof(bufferSize));
+   int connectResponse = {0};
+   std::string connectionType("");
+   if (shouldConnect){
+      connectResponse = nn_connect(mSocket,
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_connect";
+   } else{
+      connectResponse = nn_bind(mSocket,
+                                mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_bind";
+   }
+   if (connectResponse < 0) {
+      throw std::runtime_error("could not connect to endpoint: " +
+                               mProtocolHandler->GetLocation() +
+                               " because of error: " +
+                               std::string(nn_strerror(errno)));
+   }
+   LOG(INFO) << connectionType
+              << " socket at: " << mProtocolHandler->GetLocation()
+              << " with timeout: " << timeoutInMs;
+}
+/**
+ * Return the socket location
+ */
+std::string Pull::GetBinding() const {
+   return mProtocolHandler->GetLocation();
+}
+/**
+ * Receive a msg from the nanomsg queue, check for errors
+ */
+int Pull::ReceiveMsg(void *& buf, bool dontWait) {
```

Comment:
@john-gress @KjellKod yeah i'll need an opinion on this. i have no preference and it can be set in the constructor or the send/receive functions. for nanomsg the option is included in the send/receive commands, so putting it in the constructor will lock the class to that option instead of letting it change between firings. not sure if that is desired or not.